turbopack/crates/turbopack-ecmascript/src/lib.rs

```diff
@@ -968,15 +1134,301 @@ impl EcmascriptModuleContent {
+                    .current_module_contexts
+                    .iter()
+                    .find(|(_, module_ctxt)| *ctxt == **module_ctxt)
+                    .unwrap()
+                    .0
+            } else {
+                self.current_module
+            };
+
+            *ctxt = *self.merged_ctxts.get(&module).unwrap();
+        }
+        // fn visit_mut_span(&mut self, span: &mut Span) {}
+    }
+
+    let prepare_module = |(module, module_contexts, export_mark, content): &mut (
+        ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>,
+        FxIndexMap<ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>, SyntaxContext>,
+        Mark,
+        CodeGenResult,
+    )| {
+        if let CodeGenResult {
+            program: Program::Module(content),
+            globals,
+            ..
+        } = content
+        {
+            GLOBALS.set(&*globals, || {
+                content.visit_mut_with(&mut SetSyntaxContextVisitor {
+                    current_module: *module,
+                    export_mark: *export_mark,
+                    merged_ctxts,
+                    current_module_contexts: module_contexts,
+                });
+            });
+
+            content.take().body
+        } else {
+            unreachable!()
+        }
+    };
+
+    let mut inserted = FxHashSet::with_capacity_and_hasher(contents.len(), Default::default());
+
+    inserted.extend(entries.iter().map(|(_, i)| *i));
+
+    let mut merged_ast = swc_core::ecma::ast::Module {
+        span: DUMMY_SP,
+        shebang: None,
+        body: entries
+            .iter()
+            .flat_map(|(_, i)| prepare_module(&mut contents[*i]))
+            .collect(),
+    };
+
+    // Replace inserted `__turbopack_merged_esm__(i);` statements with the corresponding ith-module
+    let mut i = 0;
+    loop {
+        if i >= merged_ast.body.len() {
+            break;
+        }
+        if let ModuleItem::Stmt(Stmt::Expr(ExprStmt { expr, .. })) = &merged_ast.body[i] {
+            if let Expr::Call(CallExpr {
+                callee: Callee::Expr(callee),
+                args,
+                ..
+            }) = &**expr
+            {
+                if callee.is_ident_ref_to("__turbopack_merged_esm__") {
+                    let index = args[0].expr.as_lit().unwrap().as_num().unwrap().value as usize;
+
+                    if inserted.insert(index) {
+                        let module = &mut contents[index];
+                        merged_ast.body.splice(i..=i, prepare_module(module));
+
+                        // Don't increment, the ith item has just changed
+                        continue;
+                    } else {
+                        // Already inserted (and thus already executed), remove the placeholder
+                        merged_ast.body[i] =
+                            ModuleItem::Stmt(Stmt::Empty(EmptyStmt { span: DUMMY_SP }));
+                    }
+                }
+            }
+        }
+
+        i += 1;
+    }
```

Comment:
The splice is a bit inefficent. You can simplify that with `merged_ast.body = merged_ast.body.take().into_iter().flat_map(...)`.